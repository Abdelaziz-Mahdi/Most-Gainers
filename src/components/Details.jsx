import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useParams } from 'react-router-dom';
import {
  selectGainers, selectGainersLoading, selectGainersError,
} from '../redux/gainersSlice';
import '../styles/Details.css';

export default function Details() {
  const { symbol } = useParams();
  const dispatch = useDispatch();
  const gainers = useSelector(selectGainers);
  const detailsLoading = useSelector(selectGainersLoading);
  const detailsError = useSelector(selectGainersError);
  const [cardResults, setCardResults] = React.useState([]);
  React.useEffect(() => {
    const results = gainers.filter((gainer) => gainer.symbol.includes(symbol));
    setCardResults(results[0]);
  }, [gainers, symbol]);

  useEffect(() => {
    dispatch({ type: 'details/fetchDetails', payload: symbol });
  }, [dispatch, symbol]);

  let content;

  if (detailsLoading) {
    content = <div className="spinner-border text-success" role="status" />;
  } else if (detailsError) {
    content = <div className="alert alert-danger" role="alert">{detailsError}</div>;
  } else {
    content = (
      <div className="details-list">
        <h3 className="text-white">{cardResults.name}</h3>
        <div className="text-white">
          <p>
            Change:&nbsp;
            {cardResults.change}
            <br />
            <br />
            Change Percent:&nbsp;
            {cardResults.changesPercentage}
            <br />
            <br />
            Price:&nbsp;
            {cardResults.price}
            <br />
            <br />
            Symbol:&nbsp;
            {cardResults.symbol}
          </p>
        </div>
      </div>
    );
  }

  return (
    <section>
      <div className="details-container">
        {content}
      </div>
    </section>
  );
}
