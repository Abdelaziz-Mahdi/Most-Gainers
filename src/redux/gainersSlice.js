import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const API_URL = 'https://financialmodelingprep.com/api/v3/stock_market/gainers?apikey=3a50c9290eb377d677aba64d1291708d';

const initialState = {
  gainers: [],
  gainersLoading: false,
  gainersError: null,
};

export const fetchGainers = createAsyncThunk(
  'gainers/fetchGainers',
  async () => {
    try {
      const response = await axios.get(API_URL);
      return response.data;
    } catch (error) {
      return error;
    }
  },
);

const gainersSlice = createSlice({
  name: 'gainers',
  initialState,
  reducers: {
    filterGainers:
    (state, action) => state.gainers.filter((gainer) => gainer.name !== action.payload),
  },
  extraReducers(builder) {
    builder
      .addCase(fetchGainers.pending, (state) => {
        state.gainersLoading = true;
      })
      .addCase(fetchGainers.fulfilled, (state, action) => {
        state.gainersLoading = false;
        state.gainers = action.payload;
      })
      .addCase(fetchGainers.rejected, (state, action) => {
        state.gainersLoading = 'failed';
        state.gainersError = action.payload;
      });
  },
});

export const selectGainers = (state) => state.gainers.gainers;
export const selectGainersLoading = (state) => state.gainers.gainersLoading;
export const selectGainersError = (state) => state.gainers.gainersError;
export const { addABook, removeABook } = gainersSlice.actions;
export default gainersSlice.reducer;
