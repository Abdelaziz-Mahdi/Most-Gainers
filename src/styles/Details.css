.details-container,
.details-list {
  display: flex;
  flex-direction: column;
}

.details-container div {
  display: flex;
  padding: 20px 0 20px 0;
  justify-content: center;
  align-items: center;
  gap: 2rem;
}

ul {
  list-style: none;
  padding: 0;
  margin: 0;
  width: 100%;
}

.details-list li {
  padding: 1rem 0.5rem;
}

.details-list li:nth-child(odd) {
  background-color: rgb(204, 78, 137);
}

p {
  font-size: 1.1rem;
}

.text-white {
  color: #fff;
}

.info-text {
  color: #fff;
  font-size: 1.1rem;
}
