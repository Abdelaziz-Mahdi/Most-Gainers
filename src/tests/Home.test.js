import React from 'react';
import renderer from 'react-test-renderer';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import Details from '../components/Details';
import gainersSlice from '../redux/gainersSlice';

const store = configureStore({
  reducer: {
    gainers: gainersSlice,
  },
});

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Details />
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
