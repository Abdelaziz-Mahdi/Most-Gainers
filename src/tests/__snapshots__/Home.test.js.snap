// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section>
  <div
    className="details-container"
  >
    <div
      className="details-list"
    >
      <h3
        className="text-white"
      />
      <div
        className="text-white"
      >
        <p>
          Change: 
          <br />
          <br />
          Change Percent: 
          <br />
          <br />
          Price: 
          <br />
          <br />
          Symbol: 
        </p>
      </div>
    </div>
  </div>
</section>
`;
